### PR TITLE
start a tool to validate aspects of the blockchain database

### PR DESCRIPTION
**Pull request by arvidn**
add tool to validate some aspects of a blockchain database file. Most importantly, that the coin set is consistent with all the blocks.

To use this program:

```
cd crates/chia-tools
cargo run --release --bin validate-blockchain-db -- ~/.chia/mainnet/db/blockchain_v2_mainnet.sqlite
cargo run --release --bin validate-blockchain-db -- ~/.chia/mainnet/db/blockchain_v2_testnet11.sqlite --testnet
```

### CODE REVIEW

File: crates/chia-tools/src/bin/validate-blockchain-db.rs [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+
+                assert_eq!(conditions.cost, ti.cost);
+
+                for spend in &conditions.spends {
+                    let coin_name = *spend.coin_id;
+                    assert!(removals.remove(coin_name.as_slice()),
+                        "could not find coin {coin_name} in coin_record table, which is being spent at height {height}");
+                    for add in &spend.create_coin {
+                        let new_coin_id = Coin::new(coin_name, add.puzzle_hash, add.amount).coin_id();
+                        let Some(new_db_coin) = additions.get(new_coin_id.as_slice()) else {
```
Comment (AmineKhaldi) [inline crates/chia-tools/src/bin/validate-blockchain-db.rs:228]: I would destructure this tuple if we're not going for a struct, to increase readability and avoid mistakes.

File: crates/chia-tools/src/bin/validate-blockchain-db.rs [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+                    }
+                    panic!();
+                }
+
+                // at this point, the only additions left from the DB should be
+                // rewards. Ensure that's the case
+                let rewards = block.get_included_reward_coins();
+                for add in &rewards {
+                    let new_coin_id = add.coin_id();
+                    let Some(new_db_coin) = additions.get(new_coin_id.as_slice()) else {
```
Comment (AmineKhaldi) [inline crates/chia-tools/src/bin/validate-blockchain-db.rs:252]: Same here (destructuring the tuple).

File: crates/chia-tools/src/bin/validate-blockchain-db.rs [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+    println!("iterating over blocks starting at height {}", args.start);
+    iterate_blocks(&args.file, args.start, None, |height, block, block_refs| {
+        // If we don't start validation from height 0, we need to initialize the
+        // expected prev-hash based on the first block we pull from the DB
+        if args.start != 0 && prev_hash == constants.genesis_challenge {
+            prev_hash = block.prev_header_hash();
+        }
+
+        // this is not a transaction block
```
Comment (arvidn) [inline crates/chia-tools/src/bin/validate-blockchain-db.rs:144]: I'll move this comment to the correct place

**Review by arvidn — COMMENTED** [review]
I believe there's one case that's not checked correctly. We don't necessarily verify that unspent coins are marked as unspent.
Maybe we should ensure the coin table is empty for all non-tx blocks

File: crates/chia-tools/src/bin/analyze-chain.rs [hunk 4]
```diff
@@ -42,11 +42,14 @@ fn main() {
     let mut a = Allocator::new_limited(500_000_000);
     let allocator_checkpoint = a.checkpoint();
     let mut prev_timestamp = 0;
-    iterate_tx_blocks(
+    iterate_blocks(
         &args.file,
         args.start,
         Some(args.end),
         |height, block, block_refs| {
+            if block.transactions_generator.is_none() {
```
Comment (AmineKhaldi) [inline crates/chia-tools/src/bin/analyze-chain.rs:50]: What's the difference in all these, between `block.transactions_generator.is_none()` and `block.is_transaction_block()` aka `block.foliage.foliage_transaction_block_hash.is_some()` ?

File: crates/chia-tools/src/bin/validate-blockchain-db.rs (hunk 3 above)
Comment (AmineKhaldi) [inline crates/chia-tools/src/bin/validate-blockchain-db.rs:144]: This looks out of place, is it meant to be inside ```if block.transactions_generator.is_none() {
            return;
        }``` below?

File: crates/chia-tools/src/bin/validate-blockchain-db.rs (hunk 1 above)
Comment (AmineKhaldi) [reply crates/chia-tools/src/bin/validate-blockchain-db.rs:228]:
```suggestion
                        let Some(&(new_db_coin_ph, new_db_coin_parent, new_db_coin_amount, new_db_coin_is_reward )) = additions.get(new_coin_id.as_slice()) else {
```

File: crates/chia-tools/src/bin/analyze-chain.rs (hunk 4 above)
Comment (arvidn) [reply crates/chia-tools/src/bin/analyze-chain.rs:50]: my understanding is that they are all either supposed to be `Some` or `None`.

File: crates/chia-tools/src/bin/validate-blockchain-db.rs [hunk 5]
```diff
@@ -183,6 +183,54 @@ features that are validated:
             let amount = u64::from_be_bytes(row.get::<_, [u8; 8]>(4).expect("missing amount"));
             additions.insert(coin_name, (ph, parent, amount, reward));
         }
+
+        // first ensure that the reward coins for this block are all included in
+        // the coin record table.
+        let rewards = block.get_included_reward_coins();
+        for add in &rewards {
+            let new_coin_id = add.coin_id();
+            let Some((ph, _parent, amount, coin_base)) = additions.get(new_coin_id.as_slice())
```
Comment (AmineKhaldi) [inline crates/chia-tools/src/bin/validate-blockchain-db.rs:192]:
```suggestion
            let Some(&(ph, _parent, amount, coin_base)) = additions.get(new_coin_id.as_slice())
```
Leveraging the reference we get from `get` allow us to compare `amount` and `coin_base` values directly.

File: crates/chia-tools/src/bin/validate-blockchain-db.rs [hunk 6]
```diff
@@ -183,6 +183,54 @@ features that are validated:
...
+            // TODO: ensure the parent coin ID is set correctly
+            assert_eq!(ph, add.puzzle_hash.as_slice(),
+                "at height {height} the reward coin {new_coin_id} has an incorrect puzzle hash in the coin_record table {} expected {}",
+                hex::encode(ph),
+                add.puzzle_hash
+            );
+            // ensure the parent hash has the expected look
+            assert_eq!(*amount, add.amount, "at height {height} reward coin {new_coin_id} has amount {} in coin_record table, but the block has amount {}", amount, add.amount);
+            // this is a reward coin
+            assert!(coin_base, "at height {height} the reward coin {new_coin_id} is not marked as coin-base in the database");
```
Comment (AmineKhaldi) [inline crates/chia-tools/src/bin/validate-blockchain-db.rs:205]: This doesn't look right, IIUC we're checking a reference to bool, not bool itself. Applying the suggestion above would make this comparison againt the bool directly.

File: crates/chia-tools/src/bin/validate-blockchain-db.rs [hunk 7]
```diff
@@ -225,14 +273,14 @@ features that are validated:
                         "could not find coin {coin_name} in coin_record table, which is being spent at height {height}");
                     for add in &spend.create_coin {
                         let new_coin_id = Coin::new(coin_name, add.puzzle_hash, add.amount).coin_id();
-                        let Some(new_db_coin) = additions.get(new_coin_id.as_slice()) else {
+                        let Some((ph, parent, amount, coin_base)) = additions.get(new_coin_id.as_slice()) else {
```
Comment (AmineKhaldi) [inline crates/chia-tools/src/bin/validate-blockchain-db.rs:276]: Same notes above apply to this one and below.

**Review by AmineKhaldi — APPROVED** [review]

**Pull request merged**